### ISSUE TITLE
RFC: Alternative strategy for network defaulting

### ISSUE THREAD

**Issue by hoegaarden** [kubernetes-sigs/kubeadm-dind-cluster#203]
### Background

In #161 the option to run multiple clusters in parallel was introduced. Users needed to explicitly configure the subnet & external port for the kube-apiserver for every but the first cluster. To have a better UX #166 introduced mechanisms to find a free subnet for a cluster and a free port to expose the kube-apiserver on. There are some problems with the current mechanism how free networks are being found, which is why we'd like to kick of a discussion here on how other people think this problem should be solved. Sidenote: We didn't get any reports on issues with the defaulting for the kube-apiserver port, thus this mechanism will stay as is and is not part of this discussion.

### Current Mechanism

- If there is a docker network for a cluster, `DIND_SUBNET` and `DIND_SUBNET_SIZE` is queried from that docker network -> **done**
- If there is no docker network for a cluster
  - if the user has not specifically set `DIND_SUBNET_SIZE` it will be defaulted to `16`
  - if the user has not specifically set `DIND_SUBNET` it well be defaulted:
    - all docker networks are checked and the lowest IP is determined after which a network of size `DIND_SUBNET_SIZE` can be placed -> **done**
   Note: different subnet size of already existing docker networks and of the `DIND_SUBNET_SIZE` are considered and handled correctly.
  - If a user has set `DIND_SUBNET` / `DIND_SUBNET_SIZE` those settings are used -> **done**

Notes:
- With different network size (eg. `/16` vs. `/24` vs. `/18`) in play the and the current implementation the assumption that a network spans from `x.x.x.0` to `x.x.x.255` or `x.x.0.0` to `x.x.255.255` is not true anymore. Therefore getting the *n*th IP of a network cannot be done by string concatenation anymore but the IP has to actually be calculated.
- To minimise dependencies, currently no external tools are in play (e.g. `ipcalc`) but all calculation is done in shell.

### Current Problems

- This is only available for IPv4 currently. This is (the main) reason why currently multiple clusters for IPv6 are specifically disabled.
- It is relatively complicated to do IP arithmetics in shell, even more so when IPv6 comes in.
- Only the docker networks are considered. If users have any other networks on their machines, collisions might still occur.

### Suggestions

1. Fix current implementation

    Description:
    - Fix potential collisions by also excluding network (and other stuff like routes?) on the host
    - Add same mechanics for IPv6:
      - discover where a network of a certain size can be placed
      - implement the method to get the *n*th IP of a IPv6 network

    Pros:
    - large number of nodes are supported ([issue](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/issues/170))
    - works on any system, regardless of already allocated networks
    - works with any netmask for already allocated networks and for the network to be claimed

    Cons:
    - Lots of rather complicated code for IP arithmetic (esp. when we stick with no external dependencies)
    - Unpredictable networks, depending on state of a user's machine

1. Fixed block of networks k-d-c picks from and simple numbering of docker resources including networks

    Description:
    - rollback #166 (at least the network discovery part)
    - fix `DIND_SUBNET_SIZE` (e.g. `/24` for IPv4)
    - consider `10.192.x.0-255` ranges for k-d-c clusters (same mechanism for IPv6)
    - maybe don't use the hash of `DIND_LABEL` to name docker networks (and other resources) but use a counter, e.g. `DIND_NR`
      - first cluster: network name = "kubeam-dind-net", master name = "kube-master", subnet = 10.192.**0**.0-255
      - 2nd cluster: network name = "kubeadm-dind-net-**1**", master name = "kube-master-**1**", subnet = 10.192.**1**.0-255
      - 3nd cluster: network name = "kubeadm-dind-net-**2**", master name = "kube-master-**2**", subnet = 10.192.**2**.0-255

    Pros:
    - easy to implement
    - easy to understand
    - no complicated IP arithmetic

    Cons:
    - more potential clashes with users local setup
    - fixed subnet size
    - there should be a mechanism to create a random second cluster on a free subnet ; so there needs to be a mechanism on how to target that cluster potentially without knowing the `DIND_NR` up front.

### General Questions

- Is it ok to remove the option to configure the `DIND_SUBNET_SIZE`?

----

- Many of the above is based on input from @ivan4th & @pmichali (thanks!) -- In case I have misunderstood (parts of) your suggestions please let me know!
- In case other suggestions come in (which I would highly appreciate!) I will add them here to that issue.
- I hope to reach consensus here in *x* weeks (TBD, I suggest 2 weeks) and then generate a new issue with the agreed upon mechanism outlined (or multiple issue) -- hopefully we can then close those with PRs implementing whatever we came up with.
- I will try to keep the issue up to date when comments come in, however if people prefer to move this to a google doc or somewhere where people can actually edit things, I am happy to do so!

@pmichali 
@ivan4th 
@mariantalla

**Comment by pmichali** [comment]
Another option to consider...

Allow the user to specify a cluster CIDR, from which networks would be allocated programatically from for each cluster. If not specified, a default could be used. This way, instead fo trying to find a free network, the user can easily specify a non-colliding network.

From the cluster CIDR, a fixed size could be used for each network inside that CIDR.

For example, for IPv4:

Cluster CIDR=10.1.0.0/16
Cluster #1 uses 10.1.1.0/24
Cluster #2 uses 10.1.2.0/24
...
Cluster #N uses 10.1.N.0/24

Nodes in Cluster #1 use 10.1.1.X/24, where X is from 1..253. Alternatively, Cluster CIDR could be /8 and cluster networks using /16.

For IPv6, which is easier:

Cluster CIDR=fd00:10:20:30::/64
Cluster #1 uses fd00:10:20:30:100::/72
Cluster #2 uses fd00:10:20:30:200::/72
...
Cluster #N uses fd00:10:20:30:N00::/72

Nodes in Cluster #1 uses fd00:10:20:30:1XX::/72, where X is 0x01,0x02,0x03,... for each node.

One could also specify a size for the cluster CIDR, to allow the user the ability to set the size of the networks. So, if a user wants networks that are /64, they could specify a Cluster CIDR prefix of /56. For example, fd00:10::/56, giving cluster networks fd00:10:0:1::/64, fd00:10:0:2::/64,...

This is already in place for IPv6 for pod networks, where sizes are set by a prefix and size and then subnets calculated to be of size N+8.

The nice thing here, is that a cluster ID can be defined, going from 1..N, and number of nodes from 1(or 3)..M, and addresses can be easily calculated. It avoids algorithms altogethers.

We likely won't have more than 256 clusters and 256 nodes per cluster.
